### PR TITLE
Fully support mixed instance policy

### LINKED ISSUE #231: ec2_asg mixed_instance_policy does not support instances_distribution
##### SUMMARY
ec2_asg mixed_instance_policy does not support instances_distribution. It only supports setting multiple instance types. Instead, it should support all of the features in https://docs.aws.amazon.com/autoscaling/ec2/APIReference/API_MixedInstancesPolicy.html, not just instance type overrides.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
ec2_asg

##### ADDITIONAL INFORMATION
This would be useful for users trying to add spot instances to their ASGs.

### PR DESCRIPTION
**Pull request by JacobHenner**
##### SUMMARY

Previously, setting instances_distribution was not supported.
instances_distribution should be supported, to allow users to enable
spot instances within their mixed instance ASGs.

Note: The type and significance of the mixed_instance_policy has
changed. It now captures all of the mixed_instance_policy configuration
parameters, rather than just a list of instance types.

Fixes #231

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
ec2_asg

### CODE REVIEW

**Comment by JacobHenner** [comment]
I just realized tests were not updated - handling that now.

**Comment by JacobHenner** [comment]
It looks like some shippable tests have failed due to issues pulling docker images.

**Review by tremble — CHANGES_REQUESTED** [review]
Thanks for taking the time to submit this change.  I've retriggered the failed CI tests and they're now passing.

Mostly nits:

- Please add some additional tasks to the integration tests rather than updating the existing tasks: it's important that we know that the existing behaviour doesn't change and break as people move between versions.
- Please add a changelog fragment: https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to
- Unfortunately I think we need to use a different name for the new return data, switching from a list to a dict will break existing playbooks.  (full_mixed_instances_policy ?)  (@jillr @garethr any opinions?)

**Comment by JacobHenner** [comment]
Thanks for the feedback. I'll make the requested adjustments either later today or tomorrow.

File: plugins/modules/ec2_asg.py [hunk 1]
```diff
@@ -448,10 +505,32 @@
     type: int
     sample: 1
 mixed_instance_policy:
-    description: Returns the list of instance types if a mixed instance policy is set.
+    description: Returns the dictionary representation of the mixed instance policy.
     returned: success
-    type: list
-    sample: ["t3.micro", "t3a.micro"]
+    type: dict
```
[the remark this answers is not in the document]
Comment (JacobHenner) [reply plugins/modules/ec2_asg.py]: Changes made in response to feedback

**Review by markuman — APPROVED** [review]
LGTM

**Comment by markuman** [comment]
@JacobHenner can you resolve the change requst by @tremble?

File: plugins/modules/ec2_asg.py [hunk 2]
```diff
@@ -805,6 +892,14 @@ def get_launch_object(connection, ec2_connection):
                 for instance_type in instance_types:
                     instance_type_dict = {'InstanceType': instance_type}
                     policy['LaunchTemplate']['Overrides'].append(instance_type_dict)
+            if instances_distribution:
+                instances_distribution_params = dict(
```
Comment (tremble) [inline plugins/modules/ec2_asg.py:896]: I personally find this flow/formatting a little harder to follow than it should be.

Tests show that the code works, and I know that this is a valid pythonic way to do this.  As such I'll put my +1 but I'd like to have @jillr's opinion if there's a better way to format this before we merge

**Review by tremble — APPROVED** [review]
The code looks sound.  I'd like @jillr or @s-hertel to have a quick look and do the merge if they're happy with the formatting.

**Review by jillr — COMMENTED** [review]
The code looks good but I haven't been able to test successfully, #230 is breaking the test suite and I haven't sorted that out yet. @tremble were you able to test this change?

**Comment by goneri** [comment]
@JacobHenner, don't be surprised, I rebased your PR :-).

**Review by jillr — COMMENTED** [review]
Apologies this one stalled @JacobHenner, I think we just need to add back the import for camel_dict_to_snake_dict (I suspect it was removed in another commit now that this PR has been rebased). That should get the sanity tests passing so that we can merge.

**Review by tremble — CHANGES_REQUESTED** [review]
A few minor bits and pieces.  Let's see if we can get this over the finishing line

**Review by tremble — APPROVED** [review]
Ran the tests locally.  ec2_launch_template is currently broken but once fixed the unsupported ec2_asg tests pass with this change.

File: plugins/modules/ec2_asg.py [hunk 3]
```diff
@@ -93,6 +93,60 @@
...
+        required: false
+      instances_distribution:
+        description:
+          - >-
+            Specifies the distribution of On-Demand Instances and Spot Instances, the maximum price
+            to pay for Spot Instances, and how the Auto Scaling group allocates instance types
+            to fulfill On-Demand and Spot capacity.
+          - 'See also U(https://docs.aws.amazon.com/autoscaling/ec2/APIReference/API_InstancesDistribution.html)'
+        required: false
+        type: dict
```
Comment (tremble) [inline plugins/modules/ec2_asg.py:105]:
```suggestion
        type: dict
        version_added: 1.5.0
```

File: plugins/modules/ec2_asg.py [hunk 4]
```diff
@@ -93,6 +93,60 @@
...
+        suboptions:
+          on_demand_allocation_strategy:
+            description:
+              - Indicates how to allocate instance types to fulfill On-Demand capacity.
+            type: str
+            required: false
```
Comment (tremble) [inline plugins/modules/ec2_asg.py:111]:
```suggestion
            required: false
            version_added: 1.5.0
```

File: plugins/modules/ec2_asg.py [hunk 5]
```diff
@@ -93,6 +93,60 @@
...
+          on_demand_base_capacity:
+            description:
+              - >-
+                The minimum amount of the Auto Scaling group's capacity that must be fulfilled by On-Demand
+                Instances. This base portion is provisioned first as your group scales.
+              - >-
+                Default if not set is 0. If you leave it set to 0, On-Demand Instances are launched as a
+                percentage of the Auto Scaling group's desired capacity, per the OnDemandPercentageAboveBaseCapacity setting.
+            type: int
+            required: false
```
Comment (tremble) [inline plugins/modules/ec2_asg.py:121]:
```suggestion
            required: false
            version_added: 1.5.0
```

File: plugins/modules/ec2_asg.py [hunk 6]
```diff
@@ -93,6 +93,60 @@
...
+          on_demand_percentage_above_base_capacity:
+            description:
+              - Controls the percentages of On-Demand Instances and Spot Instances for your additional capacity beyond OnDemandBaseCapacity.
+              - Default if not set is 100. If you leave it set to 100, the percentages are 100% for On-Demand Instances and 0% for Spot Instances.
+              - 'Valid range: 0 to 100'
+            type: int
+            required: false
```
Comment (tremble) [inline plugins/modules/ec2_asg.py:128]:
```suggestion
            required: false
            version_added: 1.5.0
```

File: plugins/modules/ec2_asg.py [hunk 7]
```diff
@@ -93,6 +93,60 @@
...
+          spot_allocation_strategy:
+            description:
+              - Indicates how to allocate instances across Spot Instance pools.
+            type: str
+            required: false
```
Comment (tremble) [inline plugins/modules/ec2_asg.py:133]:
```suggestion
            required: false
            version_added: 1.5.0
```

File: plugins/modules/ec2_asg.py [hunk 8]
```diff
@@ -93,6 +93,60 @@
...
+          spot_instance_pools:
+            description:
+              - >-
+                The number of Spot Instance pools across which to allocate your Spot Instances. The Spot pools are determined from
+                the different instance types in the Overrides array of LaunchTemplate. Default if not set is 2.
+              - Used only when the Spot allocation strategy is lowest-price.
+              - 'Valid Range: Minimum value of 1. Maximum value of 20.'
+            type: int
+            required: false
```
Comment (tremble) [inline plugins/modules/ec2_asg.py:142]:
```suggestion
            required: false
            version_added: 1.5.0
```

File: plugins/modules/ec2_asg.py [hunk 9]
```diff
@@ -93,6 +93,60 @@
...
+          spot_max_price:
+            description:
+              - The maximum price per unit hour that you are willing to pay for a Spot Instance.
+              - If you leave the value of this parameter blank (which is the default), the maximum Spot price is set at the On-Demand price.
+              - To remove a value that you previously set, include the parameter but leave the value blank.
+            type: str
+            required: false
```
Comment (tremble) [inline plugins/modules/ec2_asg.py:149]:
```suggestion
            required: false
            version_added: 1.5.0
```

**Review by tremble — APPROVED** [review]
Noticed some missing version_added.

**Pull request merged**